### PR TITLE
Modify KNIGHT baseline to fuse2 style

### PR DESCRIPTION
**Pull request by liamhazan**
Modified KNIGHT example to fuse2 style 
added augmentation OP

### CODE REVIEW

**Review by alex-golts — COMMENTED** [review]
Thanks Liam, 
please see some comments that I had. we can go over some of them face to face

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 1]
```diff
@@ -1,198 +1,264 @@
...
+            row['radiographic_size'] = 4.1 # this is the median value on the training set
+
+        sample_dict["data.gt.gt_global.task_1_label"] = int(row["aua_risk_group"] in ['high_risk', 'very_high_risk'])
+        sample_dict["data.gt.gt_global.task_2_label"] = ['benign','low_risk','intermediate_risk','high_risk', 'very_high_risk'].index(row["aua_risk_group"])
+
+        sample_dict["data.input.clinical"] = row
+
+        return sample_dict
+        
+class OpPrepare_Clinical(OpBase):
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:91]: Try to implement generic building blocks that can be reused in different datasets.
Use OpLookup (you might need minor modification) or implement a new op to handle missing values.
Use OpToRange to move it a range between 0 and 1.
Use OpToOneHot to convert to one-hot vector.
And OpConcat to concatenate.

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 2]
```diff
@@ -1,198 +1,264 @@
...
+        (OpKnightSampleIDDecode(), dict()), # will save image and seg path to "data.input.img_path", "data.gt.seg_path" and load json data
+        (OpClinicalLoad(os.path.join(data_dir, 'knight.json')), dict()),
+        # loading data
+        (OpLoadImage(data_dir), dict(key_in="data.input.img_path", key_out="data.input.img", format="nib")),
+        # (OpLoadImage(data_dir), dict(key_in="data.gt.seg_path", key_out="data.gt.seg", format="nib")),
+        
+        
+        # fixed image normalization
+        (OpClip(), dict(key="data.input.img", clip=(-62, 301))),
+        (OpLambda(lambda x: (x - 104.0)/75.3 ), dict(key="data.input.img")), #kits normalization
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:155]: Can you implement OpZScoreNorm for this one (seems like a useful op)

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 3]
```diff
@@ -1,198 +1,264 @@
...
+        # affine transformation per slice but with the same arguments
+        (OpRandApply(OpSample(OpAugAffine2D()), 0.5) , dict(
+            key="data.input.img",
+            rotate=Uniform(-180.0,180.0),
+            scale=Uniform(0.8, 1.2),
+            flip=(RandBool(0.5).sample(), RandBool(0.5)),
+            translate=(RandInt(-15, 15).sample(), RandInt(-15, 15))
+        )),
+        (OpRandomCrop(), dict(key="data.input.img", out_size=resize_to)),
+        # (OpLambda(lambda x: aug_op_random_crop_and_pad(x, resize_to, centralize=False)), dict(key="data.input.img")),
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:198]: remove commented out code (in all other places as well)

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 4]
```diff
@@ -1,198 +1,264 @@
...
         print(f'- Load and cache data:')
-        train_dataset.create(reset_cache=reset_cache)
+        train_dataset.create()
     
         print(f'- Load and cache data: Done')
 
         ## Create sampler
         print(f'- Create sampler:')
-        sampler = SamplerBalancedBatch(dataset=train_dataset,
+        sampler = BatchSamplerDefault(dataset=train_dataset,
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:232]: The splitting and balancing should move to be part of the training scheme.

File: examples/fuse_examples/imaging/classification/knight/baseline/fuse_baseline.py [hunk 5]
```diff
@@ -1,5 +1,6 @@
 from collections import OrderedDict
 import pathlib
+from pickle import TRUE
```
Comment (mosheraboh) [inline examples/fuse_examples/imaging/classification/knight/baseline/fuse_baseline.py:3]: ?

**Review by mosheraboh — COMMENTED** [review]
Thanks a lot Liam!
Please find my comments inline.
And feel free to reach out if the comments are not clear and you prefer to talk.

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 6]
```diff
@@ -1,198 +1,264 @@
...
+        
+        '''
+
+        sid = get_sample_id(sample_dict)
+        
+        img_filename_key = 'data.input.img_path'
+        sample_dict[img_filename_key] =   os.path.join(sid, 'imaging.nii.gz')
+
+        seg_filename_key = 'data.gt.seg_path'
+        sample_dict[seg_filename_key] = os.path.join(sid, 'aggregated_MAJ_seg.nii.gz')
```
[the remark this answers is not in the document]
Comment (liamhazan) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:53]: it is for later.

File: examples/fuse_examples/imaging/classification/knight/baseline/fuse_baseline.py [hunk 7]
```diff
@@ -30,15 +31,15 @@
...
+force_gpus = [experiment_num*2,(experiment_num*2)+1] # specify the GPU indices you want to use
+use_data = {'imaging': True, 'clinical': False} # specify whether to use imaging, clinical data or both
 batch_size = 2
-resize_to = (256, 256, 110) 
-print_and_visualize = True
+resize_to = (110, 256, 256) 
 
 if task_num == 1:
-    num_epochs = 100
+    num_epochs = 150
```
[the remark this answers is not in the document]
Comment (liamhazan) [reply examples/fuse_examples/imaging/classification/knight/baseline/fuse_baseline.py:42]: Yes, it is needed, to reach 100 accuracy on train we need even more and usually it helps performance on validation too.

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 8]
```diff
@@ -1,198 +1,264 @@
...
+        return sample_dict
+
+class OpClinicalLoad(OpBase):
+    def __init__(self, json_path: str):
+        super().__init__()
+        self.json_path = json_path
+
+    def __call__(self, sample_dict: NDict) -> NDict:
+        cols = ['case_id', 'age_at_nephrectomy', 'body_mass_index', 'gender', 'comorbidities', \
+                'aua_risk_group', 'smoking_history', 'radiographic_size', 'last_preop_egfr']
```
[the remark this answers is not in the document]
Comment (alex-golts) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py]: @liamhazan where in your code do you make use of the "knight_test_labels.json" file? I couldn't find it in dataset.py. 
for comparison: https://github.com/IBM/fuse-med-ml/blob/928375828ff321d2bf7b2084389e34e1db0682e9/fuse_examples/classification/knight/baseline/dataset.py#L118

if is required for running make_test_labels.py
in the case when you're the "oracle"/"competition organizer" who *has* the test labels and wants to evaluate results

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 9]
```diff
@@ -1,198 +1,258 @@
 
+import json
 import os
+from xmlrpc.client import Boolean
```
Comment (alex-golts) [inline examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:4]: what is this?

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 10]
```diff
@@ -1,198 +1,264 @@
...
+        (OpLambda(lambda x: x.unsqueeze(dim=0)), dict(key="data.input.img")),  
+    ]) 
+
+    train_dynamic_pipeline = PipelineDefault("dynamic", [
+                
+        # resize image to (110, 256, 256)
+        # (OpLambda(func=partial(my_resize, resize_to=(110, 256, 256))), dict(key="data.input.img")),
+
+        # Numpy to tensor
+        (OpToTensor(), dict(key="data.input.img")),
```
[the remark this answers is not in the document]
Comment (alex-golts) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:183]: @liamhazan you don't agree? I see that it wasn't changed

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 11]
```diff
@@ -1,198 +1,264 @@
...
+        (OpToTensor(), dict(key="data.input.clinical.all")),
+
+        # affine transformation per slice but with the same arguments
```
[the remark this answers is not in the document]
Comment (alex-golts) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:186]: so it's not done in the end, right? or I'm missing something?

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py [hunk 12]
```diff
@@ -1,198 +1,264 @@
...
+        (OpRandApply(OpSample(OpAugAffine2D()), 0.5) , dict(
+            key="data.input.img",
```
[the remark this answers is not in the document]
Comment (alex-golts) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py:188]: @liamhazan ?

**Review by alex-golts — COMMENTED** [review]
Thanks Liam, 
a few minor things are still not clear to me, see comments inline

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py (hunk 2 above)
Comment (liamhazan) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py]: done

File: examples/fuse_examples/imaging/classification/knight/baseline/dataset.py (hunk 3 above)
Comment (liamhazan) [reply examples/fuse_examples/imaging/classification/knight/baseline/dataset.py]: done

File: fuseimg/data/ops/shape_ops.py [hunk 13]
```diff
@@ -83,6 +84,47 @@ def __call__(self, sample_dict: NDict, op_id: Optional[str], key: str,
...
+                           centralize: bool=False
+        ):
+        aug_input = sample_dict[key]
+        assert len(aug_input.shape) == len(out_size)
+        depth, height, width = aug_input.shape #input is in the form [D,H,W]
+
+        aug_tensor = torch.full(out_size, fill, dtype=torch.float32)
+
+        if depth > out_size[0]:
+            crop_start = RandInt(0, depth - out_size[0]).sample()
```
[the remark this answers is not in the document]
Comment (liamhazan) [reply fuseimg/data/ops/shape_ops.py]: I f we don't sample inside the op it is not random crop anymore, to get random crop the user needs both crop and OpSample even though it is a basic transformation (as there is one in torchvision for the 2D case).
I can't think on a case where there is a use for such op without the randomness.

**Review by alex-golts — APPROVED** [review]

**Review by alex-golts — APPROVED** [review]
